Gm wfpr

mgalbier opened the pull request:
- Check for isEmpty() instead of isNull() to also catch empty strings
- Add option to accumulate changes for CHistory command
- Include node symbol name in object path if available
- Initial version of name change detection
- Remove unnecessary resolvePending() call
  <a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68246879%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68246922%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68247457%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68247717%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68247997%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68248127%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68248438%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68248733%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68249318%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68249603%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68249662%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68249797%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68397874%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68398102%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68398610%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23issuecomment-228348466%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23issuecomment-228350749%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23issuecomment-228353924%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23issuecomment-228348466%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22OK%2C%20that%27s%20all.%20I%20assume%20you%20have%20tested%20the%20latest%20version%3F%22%2C%20%22created_at%22%3A%20%222016-06-24T13%3A41%3A13Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22You%20assume%20correctly%21%20Thanks%22%2C%20%22created_at%22%3A%20%222016-06-24T13%3A50%3A30Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%2C%20%7B%22body%22%3A%20%22Thanks%22%2C%20%22created_at%22%3A%20%222016-06-24T14%3A03%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20f84c3627248b39de08b488e6e89dead078763aa8%20VersionControlUI/src/DiffManager.cpp%2047%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68247457%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Don%27t%20do%20the%20same%20work%20twice%3A%5Cr%5Cn%5Cr%5Cn%20%20%20%20if%28%20auto%20oldNode%20%3D%20DCast%3C...%3E%20...%29%5Cr%5Cn%20%20%20%20%20%20%20%20if%20%28auto%20newNode%20%3D%20DCast%3C...%3E...%29%5Cr%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20here%20you%20have%20access%20to%20both%20if%20they%20were%20not%20null%22%2C%20%22created_at%22%3A%20%222016-06-23T14%3A42%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/DiffManager.cpp%3AL125-211%22%7D%2C%20%22Pull%20f84c3627248b39de08b488e6e89dead078763aa8%20VersionControlUI/src/DiffManager.cpp%2058%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68247717%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22like%20above%2C%20avoid%20doing%20double%20work.%22%2C%20%22created_at%22%3A%20%222016-06-23T14%3A43%3A48Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/DiffManager.cpp%3AL125-211%22%7D%2C%20%22Pull%20f84c3627248b39de08b488e6e89dead078763aa8%20VersionControlUI/src/DiffManager.cpp%2042%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68248733%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20don%27t%20like%20this%20method%20name%20since%20it%20does%20additional%20work%20than%20just%20tell%20you%20yes/no.%20Use%20a%20more%20elaborate%20name%2C%20like%20%60processNameChange%60%22%2C%20%22created_at%22%3A%20%222016-06-23T14%3A48%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/DiffManager.cpp%3AL125-211%22%7D%2C%20%22Pull%20f84c3627248b39de08b488e6e89dead078763aa8%20VersionControlUI/src/DiffManager.cpp%2015%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68249662%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22It%20would%20be%20nice%20if%20the%20message%20has%20a%20header%20of%20some%20sort%3A%5Cr%5CnThe%20following%20objects%20have%20been%20renamed%3A%22%2C%20%22created_at%22%3A%20%222016-06-23T14%3A53%3A36Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/DiffManager.cpp%3AL125-211%22%7D%2C%20%22Pull%20f84c3627248b39de08b488e6e89dead078763aa8%20VersionControlUI/src/DiffManager.cpp%2065%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68247997%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Hmmm%2C%20what%20if%20we%20see%20this%20reference%20change%20%2A%2Abefore%2A%2A%20we%20see%20the%20NameText%20change%2C%20Would%20we%20still%20detect%20this%20renaming%3F%22%2C%20%22created_at%22%3A%20%222016-06-23T14%3A45%3A12Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22No%20in%20that%20case%20we%20would%20not%20detect%20it%20the%20way%20it%20is%20now.%20Is%20this%20possible%20to%20happen%3F%20%22%2C%20%22created_at%22%3A%20%222016-06-23T14%3A47%3A24Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%2C%20%7B%22body%22%3A%20%22Sure.%20You%27re%20just%20iterating%20over%20a%20QHash%20of%20changes%20and%20the%20order%20of%20these%20is%20totally%20arbitrary.%5Cr%5Cn%5Cr%5CnYou%20need%20to%20look%20in%20two%20stages.%20First%20Specifically%20look%20for%20changing%20NameText%2C%20and%20then%20for%20Reference.%22%2C%20%22created_at%22%3A%20%222016-06-23T14%3A51%3A50Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Hm%20very%20true%20%3A%29%20%20Will%20do.%20%22%2C%20%22created_at%22%3A%20%222016-06-23T14%3A53%3A18Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/DiffManager.cpp%3AL125-211%22%7D%2C%20%22Pull%20f84c3627248b39de08b488e6e89dead078763aa8%20VersionControlUI/src/DiffManager.cpp%2091%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68246879%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Why%20do%20we%20skip%20the%20steps%20below%20if%20this%20is%20a%20name%20change%3F%22%2C%20%22created_at%22%3A%20%222016-06-23T14%3A39%3A52Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Please%20add%20a%20comment%20here%20explaining%20that.%22%2C%20%22created_at%22%3A%20%222016-06-23T14%3A40%3A03Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22This%20avoids%20having%20highlights%20for%20name%20changes%2C%20since%20they%20are%20summarized%20in%20a%20message%20overlay%20at%20the%20moment.%20%22%2C%20%22created_at%22%3A%20%222016-06-23T14%3A45%3A48Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%2C%20%7B%22body%22%3A%20%22Change%20the%20method%20name%20and%20put%20a%20brief%20comment%20here.%22%2C%20%22created_at%22%3A%20%222016-06-23T14%3A54%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/DiffManager.cpp%3AL227-236%22%7D%2C%20%22Pull%20c92e91282a8082cbe87ed812b3897e8784abae8b%20VersionControlUI/src/DiffManager.cpp%2075%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68397874%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22You%20can%20drop%20this%20line%2C%20the%20next%20two%20cover%20this%20case%20and%20are%20just%20as%20efficient.%22%2C%20%22created_at%22%3A%20%222016-06-24T13%3A32%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ah%20I%20thought%20the%20cast%20would%20be%20more%20expensive.%20OK%20will%20do%22%2C%20%22created_at%22%3A%20%222016-06-24T13%3A33%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/DiffManager.cpp%3AL172-244%22%7D%2C%20%22Pull%20c92e91282a8082cbe87ed812b3897e8784abae8b%20VersionControlUI/src/DiffManager.cpp%2082%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/403%23discussion_r68398610%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22This%20should%20always%20exist%2C%20please%20add%20a%20Q_ASSERT.%22%2C%20%22created_at%22%3A%20%222016-06-24T13%3A37%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/DiffManager.cpp%3AL172-244%22%7D%7D%7D'></a>
  <a href='https://www.codereviewhub.com/'><img src='http://www.codereviewhub.com/site/github-bar.png' height=40></a>
- [ ] <a href='#crh-comment-Pull f84c3627248b39de08b488e6e89dead078763aa8 VersionControlUI/src/DiffManager.cpp 91'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/403#discussion_r68246879'>File: VersionControlUI/src/DiffManager.cpp:L227-236</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Why do we skip the steps below if this is a name change?
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Please add a comment here explaining that.
- <a href='https://github.com/mgalbier'><img border=0 src='https://avatars.githubusercontent.com/u/17177211?v=3' height=16 width=16'></a> This avoids having highlights for name changes, since they are summarized in a message overlay at the moment.
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Change the method name and put a brief comment here.
- [ ] <a href='#crh-comment-Pull f84c3627248b39de08b488e6e89dead078763aa8 VersionControlUI/src/DiffManager.cpp 47'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/403#discussion_r68247457'>File: VersionControlUI/src/DiffManager.cpp:L125-211</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Don't do the same work twice:
  if( auto oldNode = DCast<...> ...)
  if (auto newNode = DCast<...>...)
  here you have access to both if they were not null
- [ ] <a href='#crh-comment-Pull f84c3627248b39de08b488e6e89dead078763aa8 VersionControlUI/src/DiffManager.cpp 58'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/403#discussion_r68247717'>File: VersionControlUI/src/DiffManager.cpp:L125-211</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> like above, avoid doing double work.
- [ ] <a href='#crh-comment-Pull f84c3627248b39de08b488e6e89dead078763aa8 VersionControlUI/src/DiffManager.cpp 65'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/403#discussion_r68247997'>File: VersionControlUI/src/DiffManager.cpp:L125-211</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Hmmm, what if we see this reference change **before** we see the NameText change, Would we still detect this renaming?
- <a href='https://github.com/mgalbier'><img border=0 src='https://avatars.githubusercontent.com/u/17177211?v=3' height=16 width=16'></a> No in that case we would not detect it the way it is now. Is this possible to happen?
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Sure. You're just iterating over a QHash of changes and the order of these is totally arbitrary.
  You need to look in two stages. First Specifically look for changing NameText, and then for Reference.
- <a href='https://github.com/mgalbier'><img border=0 src='https://avatars.githubusercontent.com/u/17177211?v=3' height=16 width=16'></a> Hm very true :)  Will do.
- [ ] <a href='#crh-comment-Pull f84c3627248b39de08b488e6e89dead078763aa8 VersionControlUI/src/DiffManager.cpp 42'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/403#discussion_r68248733'>File: VersionControlUI/src/DiffManager.cpp:L125-211</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> I don't like this method name since it does additional work than just tell you yes/no. Use a more elaborate name, like `processNameChange`
- [ ] <a href='#crh-comment-Pull f84c3627248b39de08b488e6e89dead078763aa8 VersionControlUI/src/DiffManager.cpp 15'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/403#discussion_r68249662'>File: VersionControlUI/src/DiffManager.cpp:L125-211</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> It would be nice if the message has a header of some sort:
  The following objects have been renamed:
- [ ] <a href='#crh-comment-Pull c92e91282a8082cbe87ed812b3897e8784abae8b VersionControlUI/src/DiffManager.cpp 75'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/403#discussion_r68397874'>File: VersionControlUI/src/DiffManager.cpp:L172-244</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> You can drop this line, the next two cover this case and are just as efficient.
- <a href='https://github.com/mgalbier'><img border=0 src='https://avatars.githubusercontent.com/u/17177211?v=3' height=16 width=16'></a> Ah I thought the cast would be more expensive. OK will do
- [ ] <a href='#crh-comment-Pull c92e91282a8082cbe87ed812b3897e8784abae8b VersionControlUI/src/DiffManager.cpp 82'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/403#discussion_r68398610'>File: VersionControlUI/src/DiffManager.cpp:L172-244</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> This should always exist, please add a Q_ASSERT.
- [ ] <a href='#crh-comment-General Comment'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/403#issuecomment-228348466'>General Comment</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> OK, that's all. I assume you have tested the latest version?
- <a href='https://github.com/mgalbier'><img border=0 src='https://avatars.githubusercontent.com/u/17177211?v=3' height=16 width=16'></a> You assume correctly! Thanks
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Thanks

<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/403?mark_as_completed=1'><img src='http://www.codereviewhub.com/site/github-mark-as-completed.png' height=26></a>&nbsp;<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/403?approve=1'><img src='http://www.codereviewhub.com/site/github-approve.png' height=26></a>&nbsp;<a href='https://github.com/dimitar-asenov/Envision/pull/403'><img src='http://www.codereviewhub.com/site/github-refresh.png' height=26></a>
<a href='#crh-end'></a>

mgalbier committed:
Check for isEmpty() instead of isNull() to also catch empty strings

mgalbier committed:
Add option to accumulate changes for CHistory command

mgalbier committed:
Include node symbol name in object path if available

mgalbier committed:
Initial version of name change detection

mgalbier committed:
Remove unnecessary resolvePending() call

dimitar-asenov commented on VersionControlUI/src/DiffManager.cpp:
```diff
@@ -147,6 +227,9 @@ void DiffManager::computeDiff(QString oldVersion, QString newVersion, QList<Chan
 		auto oldNode = const_cast<Model::Node*>(diffSetup.oldVersionManager_->nodeIdMap().node(id));
 		auto newNode = const_cast<Model::Node*>(diffSetup.newVersionManager_->nodeIdMap().node(id));
 
+		if (isNameChange(oldNode, newNode, diffSetup))
```
Why do we skip the steps below if this is a name change?

dimitar-asenov replied:
Please add a comment here explaining that.

mgalbier replied:
This avoids having highlights for name changes, since they are summarized in a message overlay at the moment.

dimitar-asenov replied:
Change the method name and put a brief comment here.

dimitar-asenov commented on VersionControlUI/src/DiffManager.cpp:
```diff
@@ -125,6 +125,86 @@ void DiffManager::clear()
+	if (!oldNode || !newNode)
+		return false;
+
+	if (DCast<Model::NameText>(oldNode) && DCast<Model::NameText>(newNode))
```
Don't do the same work twice:

```
if( auto oldNode = DCast<...> ...)
    if (auto newNode = DCast<...>...)
            here you have access to both if they were not null
```

dimitar-asenov commented on VersionControlUI/src/DiffManager.cpp:
```diff
@@ -125,6 +125,86 @@ void DiffManager::clear()
+		return true;
+	}
+
+	if (DCast<Model::Reference>(oldNode) && DCast<Model::Reference>(newNode))
```
like above, avoid doing double work.

dimitar-asenov commented on VersionControlUI/src/DiffManager.cpp:
```diff
@@ -125,6 +125,86 @@ void DiffManager::clear()
+
+		Model::Reference::resolvePending();
+
+		auto result = nameChanges_.find(oldReference->name());
```
Hmmm, what if we see this reference change **before** we see the NameText change, Would we still detect this renaming?

mgalbier replied:
No in that case we would not detect it the way it is now. Is this possible to happen?

dimitar-asenov replied:
Sure. You're just iterating over a QHash of changes and the order of these is totally arbitrary.

You need to look in two stages. First Specifically look for changing NameText, and then for Reference.

mgalbier replied:
Hm very true :)  Will do.

dimitar-asenov commented on VersionControlUI/src/DiffManager.cpp:
```diff
@@ -125,6 +125,86 @@ void DiffManager::clear()
+}
+
+// TODO decide what should be highlighted in diff
+bool DiffManager::isNameChange(Model::Node* oldNode, Model::Node* newNode, const DiffSetup& diffSetup)
```
I don't like this method name since it does additional work than just tell you yes/no. Use a more elaborate name, like `processNameChange`

dimitar-asenov commented on VersionControlUI/src/DiffManager.cpp:
```diff
@@ -125,6 +125,86 @@ void DiffManager::clear()
+		auto oldNode = const_cast<Model::Node*>(diffSetup.oldVersionManager_->nodeIdMap().node(id));
+		auto newNode = const_cast<Model::Node*>(diffSetup.newVersionManager_->nodeIdMap().node(id));
+
+		message += DiffComparisonPair::computeObjectPath(oldNode) +
```
It would be nice if the message has a header of some sort:
The following objects have been renamed:

mgalbier committed:
Add header for name change information

mgalbier committed:
Rename isNameChange to processNameChange

mgalbier committed:
Iterate over changes in first phase to detect NameText changes

Only check for possible name changes in Reference after processing NameText changes

dimitar-asenov commented on VersionControlUI/src/DiffManager.cpp:
```diff
@@ -168,52 +172,72 @@ bool DiffManager::processNameChange(Model::Node* oldNode, Model::Node* newNode,
-		if (result != nameChanges_.end() && result->first == newReference->name())
-		{
-			if (oldReference->target() && newReference->target())
+		if (!versionNodes.oldNode_ || !versionNodes.newNode_) {iter++; continue;}
```
You can drop this line, the next two cover this case and are just as efficient.

mgalbier replied:
Ah I thought the cast would be more expensive. OK will do

dimitar-asenov commented on VersionControlUI/src/DiffManager.cpp:
```diff
@@ -168,52 +172,72 @@ bool DiffManager::processNameChange(Model::Node* oldNode, Model::Node* newNode,
 			{
-				auto oldRefId = diffSetup.oldVersionManager_->nodeIdMap().idIfExists(oldReference->target());
-				auto newRefId = diffSetup.newVersionManager_->nodeIdMap().idIfExists(newReference->target());
+				auto id = diffSetup.newVersionManager_->nodeIdMap().idIfExists(newNameText->parent());
```
This should always exist, please add a Q_ASSERT.

dimitar-asenov commented:
OK, that's all. I assume you have tested the latest version?

mgalbier commented:
You assume correctly! Thanks

mgalbier committed:
Remove unnecessary check

mgalbier committed:
Assert id exists

dimitar-asenov commented:
Thanks
